File: bionicpro-auth/src/main/java/bionicpro/auth/util/PkceUtil.java
package bionicpro.auth.util;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.util.Base64;

/**
 * PKCE (RFC 7636) — генерация code_verifier и code_challenge.
 *
 * code_verifier: 32 случайных байта → Base64URL (43 символа)
 * code_challenge: BASE64URL(SHA-256(code_verifier))
 */
public final class PkceUtil {

    private static final SecureRandom SECURE_RANDOM = new SecureRandom();

    private PkceUtil() {}

    public static String generateCodeVerifier() {
        var bytes = new byte[32];
        SECURE_RANDOM.nextBytes(bytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }

    public static String generateCodeChallenge(String codeVerifier) {
        try {
            var digest = MessageDigest.getInstance("SHA-256");
            var hash = digest.digest(codeVerifier.getBytes(java.nio.charset.StandardCharsets.US_ASCII));
            return Base64.getUrlEncoder().withoutPadding().encodeToString(hash);
        } catch (NoSuchAlgorithmException e) {
            // SHA-256 гарантированно есть в любой JVM
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    /** Генерация криптографически стойкого ID (для session_id, state). */
    public static String generateSecureId() {
        var bytes = new byte[32];
        SECURE_RANDOM.nextBytes(bytes);
        return java.util.HexFormat.of().formatHex(bytes);
    }
}

File: task 1/PKCEUtil.kt
import java.security.MessageDigest
import java.util.Base64

object PKCEUtil {
    fun generateCodeVerifier(): String {
        val byteArray = ByteArray(64)
        SecureRandom().nextBytes(byteArray)
        return Base64.getUrlEncoder().withoutPadding().encodeToString(byteArray)
    }

    fun generateCodeChallenge(codeVerifier: String): String {
        val bytes = codeVerifier.toByteArray(Charsets.US_ASCII)
        val digest = MessageDigest.getInstance("SHA-256").digest(bytes)
        return Base64.getUrlEncoder().withoutPadding().encodeToString(digest)
    }
}

// AuthRepository.kt (часть, отвечающая за PKCE)
class AuthRepository(private val httpClient: HttpClient) {
    suspend fun loginWithPKCE(username: String, password: String) {
        val codeVerifier = PKCEUtil.generateCodeVerifier()
        val codeChallenge = PKCEUtil.generateCodeChallenge(codeVerifier)

        // Сохраняем codeVerifier в защищённом хранилище (EncryptedSharedPreferences)
        saveCodeVerifier(codeVerifier)

        // 1. Запрос кода авторизации (через WebView или Custom Tabs)
        val authUrl = "${KEYCLOAK_URL}/protocol/openid-connect/auth".toHttpUrl().newBuilder()
            .addQueryParameter("response_type", "code")
            .addQueryParameter("client_id", "bionic-mobile")
            .addQueryParameter("redirect_uri", "bionic://callback")
            .addQueryParameter("code_challenge", codeChallenge)
            .addQueryParameter("code_challenge_method", "S256")
            .addQueryParameter("scope", "openid profile email")
            .build()

        // Открываем WebView для входа пользователя, перехватываем redirect_uri с кодом
        val authCode = captureAuthorizationCode(authUrl) // предполагаемая функция

        // 2. Обмен кода на токены
        val tokenResponse = httpClient.post("${KEYCLOAK_URL}/protocol/openid-connect/token") {
            setBody(
                FormDataContent(
                    listOf(
                        "grant_type" to "authorization_code",
                        "code" to authCode,
                        "client_id" to "bionic-mobile",
                        "redirect_uri" to "bionic://callback",
                        "code_verifier" to codeVerifier
                    )
                )
            )
        }
        // Обработка ответа: сохраняем access_token, refresh_token
    }
}